pyrokinetics/gk_code/GKOutputReaderCGYRO.py

```diff
@@ -0,0 +1,406 @@
+                    continue
+                raise RuntimeError(
+                    f"GKOutputReaderCGYRO: The file {cgyro_file.path.name} is needed"
+                )
+            # Read in file according to format
+            if cgyro_file.fmt == "input":
+                with open(cgyro_file.path, "r") as f:
+                    raw_data[key] = f.read()
+            if cgyro_file.fmt == "out":
+                raw_data[key] = np.loadtxt(cgyro_file.path)
+            if cgyro_file.fmt == "bin":
+                raw_data[key] = np.fromfile(cgyro_file.path, dtype="float32")
+        input_str = raw_data["input"]
+        gk_input = GKInputCGYRO()
+        gk_input.read_str(input_str)
+        return raw_data, gk_input, input_str
+
+    @classmethod
+    def _init_dataset(
+        cls, raw_data: Dict[str, Any], gk_input: GKInputCGYRO
+    ) -> xr.Dataset:
+        """
+        Sets coords and attrs of a Pyrokinetics dataset from a collection of CGYRO
+        files.
+
+        Args:
+            raw_data (Dict[str,Any]): Dict containing CGYRO output.
+            gk_input (GKInputCGYRO): Processed CGYRO input file.
+
+        Returns:
+            xr.Dataset: Dataset with coords and attrs set, but not data_vars
+        """
+
+        # Process time data
+        time = raw_data["time"][:, 0]
+
+        # Process grid data
+        grid_data = raw_data["grids"]
+        nky = int(grid_data[0])
+        nspecies = int(grid_data[1])
+        nfield = int(grid_data[2])
+        nkx = int(grid_data[3])
+        ntheta_grid = int(grid_data[4])
+        nenergy = int(grid_data[5])
+        npitch = int(grid_data[6])
+        box_size = int(grid_data[7])
+        length_x = grid_data[8]
+        ntheta_plot = int(grid_data[10])
+
+        # Iterate through grid_data in chunks, starting after kx
+        pos = 11 + nkx
+
+        theta_grid = grid_data[pos : pos + ntheta_grid]
+        pos += ntheta_grid
+
+        energy = grid_data[pos : pos + nenergy]
+        pos += nenergy
+
+        pitch = grid_data[pos : pos + npitch]
+        pos += npitch
+
+        ntheta_ballooning = ntheta_grid * int(nkx / box_size)
+        theta_ballooning = grid_data[pos : pos + ntheta_ballooning]
+        pos += ntheta_ballooning
+
+        ky = grid_data[pos : pos + nky]
+
+        if gk_input.is_linear():
+            # Convert to ballooning co-ordinate so only 1 kx
+            theta = theta_ballooning
+            ntheta = ntheta_ballooning
+            kx = [0.0]
+            nkx = 1
+        else:
+            # Output data actually given on theta_plot grid
+            ntheta = ntheta_plot
+            theta = [0.0] if ntheta == 1 else theta_grid[:: ntheta_grid // ntheta]
+            # TODO Is this correct? for nkx = 5, this linspace call gives
+            #      [-2., -1.25, -0.5, 0.25, 1.]. Was it intended  to give
+            #      [-2, -1, 0, 1, 2] instead?
+            kx = 2 * pi * np.linspace(-int(nkx / 2), int(nkx / 2) - 1, nkx) / length_x
```

Comment:
Does this need fixing? @bpatel2107

Reply:
We rarely use an odd number of `kx`, but yes good spot, it should be 


```suggestion
            # TODO Is this correct? for nkx = 5, this linspace call gives
            #      [-2., -1.25, -0.5, 0.25, 1.]. Was it intended  to give
            #      [-2, -1, 0, 1, 2] instead?
            kx = 2 * pi * np.linspace(-int(nkx / 2), int((nkx+1) / 2) - 1, nkx) / length_x
```